src/KurrentDB.SecondaryIndexing/Subscriptions/SecondaryIndexSubscription.cs

```diff
@@ -1,70 +1,84 @@
 // Copyright (c) Kurrent, Inc and/or licensed to Kurrent, Inc under one or more agreements.
 // Kurrent, Inc licenses this file to you under the Kurrent License v1 (see LICENSE.md).
 
+using System.Runtime.CompilerServices;
+using DotNext.Runtime.CompilerServices;
 using KurrentDB.Core.Bus;
 using KurrentDB.Core.Services.Storage.ReaderIndex;
 using KurrentDB.Core.Services.Transport.Common;
 using KurrentDB.Core.Services.Transport.Enumerators;
 using KurrentDB.Core.Services.UserManagement;
-using KurrentDB.SecondaryIndexing.Indices;
+using KurrentDB.SecondaryIndexing.Indexes;
 using Serilog;
 
 namespace KurrentDB.SecondaryIndexing.Subscriptions;
 
-public class SecondaryIndexSubscription(
+public sealed class SecondaryIndexSubscription(
 	IPublisher publisher,
-	ISecondaryIndex index,
-	SecondaryIndexingPluginOptions? options
+	ISecondaryIndexProcessor indexProcessor,
+	SecondaryIndexingPluginOptions options
 ) : IAsyncDisposable {
 	private static readonly ILogger Log = Serilog.Log.Logger.ForContext<SecondaryIndexSubscription>();
-	private readonly int _checkpointCommitBatchSize = options?.CheckpointCommitBatchSize ?? 50000;
-	private readonly uint _checkpointCommitDelayMs = options?.CheckpointCommitDelayMs ?? 10000;
 
-	private readonly CancellationTokenSource _cts = new();
+	private readonly int _commitBatchSize = options.CommitBatchSize;
+	private CancellationTokenSource? _cts = new();
 	private Enumerator.AllSubscription? _subscription;
 	private Task? _processingTask;
-	private SecondaryIndexCheckpointTracker? _checkpointTracker;
-
-	public async ValueTask Subscribe(CancellationToken cancellationToken) {
-		var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _cts.Token);
-		var position = await index.GetLastPosition(linkedCts.Token);
-		var startFrom = position == null ? Position.Start : Position.FromInt64((long)position, (long)position);
-
-		_checkpointTracker = new SecondaryIndexCheckpointTracker(
-			_checkpointCommitBatchSize,
-			_checkpointCommitDelayMs,
-			ct => index.Processor.Commit(ct),
-			linkedCts.Token
-		);
 
-		_subscription = new Enumerator.AllSubscription(
+	public void Subscribe() {
+		var position = indexProcessor.GetLastPosition();
+		var startFrom = position is null ? Position.Start : Position.FromInt64((long)position, (long)position);
```

Comment:
Using only the prepare position will not work properly if there are explicit transactions in the log (it might actually even move the checkpoint backwards)

It is better to use the full TFPos as checkpoint (`resolvedEvent.EventPosition!.Value.CommitPosition` instead of `resolvedEvent.Event.LogPosition`) in DefaultIndexProcessor, CategoryIndexProcessor, etc.

Reply:
(a test probably needs to be added for that case too - i.e with explicit transactions)

Reply:
I don't like that. We can read prepare records using one position. If we have a deficient enumerator that is incapable of figuring out where to start reading - that must be fixed. TFPos spreads out everywhere, including client APIs, and it just confuses people.